src/topology/mod.rs

```diff
@@ -154,34 +148,47 @@ impl RunningTopology {
         // If we reach the deadline, this future will print out which components won't
         // gracefully shutdown since we will start to forcefully shutdown the sources.
         let mut check_handles2 = check_handles.clone();
-        let timeout = delay_until(deadline).map(move |_| {
+        let timeout = async move {
+            delay_until(deadline).await;
             // Remove all tasks that have shutdown.
-            check_handles2.retain(|_name, handles| {
-                retain(handles, |handle| {
-                    handle.poll().map(|p| p.is_not_ready()).unwrap_or(false)
-                });
-                !handles.is_empty()
-            });
+            let to_check = check_handles2.drain().collect::<Vec<(String, Vec<_>)>>();
+            for (name, handles) in to_check {
+                let mut active_handles = Vec::new();
+                for handle in handles {
+                    if handle.peek().is_some() {
+                        active_handles.push(handle)
+                    }
+                }
+                if !active_handles.is_empty() {
+                    check_handles2.insert(name, active_handles);
+                }
+            }
```

Comment:
Out of interest, why drop `retain`?

```rust
check_handles2.retain(|_name, mut handles| {
  handles.retain(|handle| handle.peek().is_some())
  !handles.is_empty()
});
```
This will not work by some reason?

Reply:
Heh, good point. I originally had switched to `futures::poll!`, which must be run in an `async` block and therefore didn't work with `retain`. Now that we're using `peek` the rest of the change isn't necessary.